src/hl/container.rs

```diff
@@ -346,6 +347,115 @@ impl<'a> Writer<'a> {
     }
 }
 
+#[derive(Debug)]
+pub struct ByteReader<'a> {
+    obj: &'a Container,
+    pos: u64,
+    dt: Datatype,
+    obj_space: Dataspace,
+    xfer: PropertyList,
+}
+
+impl<'a> ByteReader<'a> {
+    pub fn new(obj: &'a Container) -> Result<Self> {
+        ensure!(!obj.is_attr(), "ByteReader cannot be used on attribute datasets");
+
+        let file_dtype = obj.dtype()?;
+        let mem_dtype = Datatype::from_type::<u8>()?;
+        file_dtype.ensure_convertible(&mem_dtype, Conversion::NoOp)?;
+
+        let obj_space = obj.space()?;
+        ensure!(obj_space.shape().len() == 1, "Only rank 1 datasets can be read vie ByteReader.");
+        let xfer = PropertyList::from_id(h5call!(H5Pcreate(*crate::globals::H5P_DATASET_XFER))?)?;
+        if !hdf5_types::USING_H5_ALLOCATOR {
+            crate::hl::plist::set_vlen_manager_libc(xfer.id())?;
+        }
+        Ok(ByteReader { obj, pos: 0, obj_space, dt: mem_dtype, xfer })
+    }
+
+    pub fn set_position(&mut self, pos: u64) {
+        self.pos = pos;
+    }
+
+    fn dataset_len(&self) -> usize {
+        self.obj_space.shape()[0]
+    }
+
+    fn remaining_len(&self) -> usize {
+        self.dataset_len().saturating_sub(self.pos as usize)
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.pos >= self.dataset_len() as u64
+    }
+}
+
+trait IOResultExt<T> {
+    fn to_io_err(self) -> std::io::Result<T>;
+}
+impl<T> IOResultExt<T> for Result<T> {
+    fn to_io_err(self) -> std::io::Result<T> {
+        match self {
+            Ok(x) => Ok(x),
+            Err(e) => Err(std::io::Error::new(std::io::ErrorKind::Other, e.to_string())),
+        }
+    }
+}
+
+impl<'a> std::io::Read for ByteReader<'a> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
```

Comment:
Not sure if it is a good idea, but this could be written
```rust
        || -> Result<usize> {
            let pos = self.pos as usize;
            let amt = std::cmp::min(buf.len(), self.remaining_len());
            let selection = Selection::new(pos..pos + amt);
            let out_shape = selection.out_shape(&self.obj_space.shape())?;
            let fspace = self.obj_space.select(selection)?;
            let mspace = Dataspace::try_new(&out_shape)?;
            h5call!(H5Dread(
                self.obj.id(),
                self.dt.id(),
                mspace.id(),
                fspace.id(),
                self.xfer.id(),
                buf.as_mut_ptr().cast()
            ))?;
            self.pos += amt as u64;
            Ok(out_shape[0])
        }()
        .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err.to_string()))
```
and we can get rid of the trait

Reply:
Hm I am not a big fan of that, but as you like. I guess I would rather use the map_err on every io::Result and not create a closure.

Reply:
1. Why can't we just implement `From<hdf5::Error> for io::Error` in `error.rs` then?
2. Why `.to_string()`? If I understand correctly, `std::io::Error` will happily box a dyn trait object

Reply:
True, I'll add it to src/error.rs